File: Algospot/nqueen.erl
-module(nqueen).
-import(lists, []).
-compile([export_all]).

vertical({_, B}, {_, D}) -> B =:= D.
diagonal({A, B}, {C, D}) ->
    DR = abs(A - C),
    DC = abs(B - D),
    case {DR div DC, DR rem DC} of
        {1, 0} ->
            true;
        _ -> false
    end.
    
safe(P, PS) ->
    not lists:any(fun (X) -> vertical(P, X) orelse diagonal(P, X) end, PS).

solve(N, PS) ->
    if length(PS) =:= N ->
            1;
       true ->
            CANDIDATES = lists:map(fun (C) -> {length(PS), C} end,
                                   lists:seq(0, N - 1)),
            lists:foldl(fun (P, Acc) -> solve(N, [P|PS]) + Acc end, 0, lists:filter(fun (P) -> safe(P, PS) end, CANDIDATES))
    end.

solve(N) ->
    solve(N, []). 

File: Algospot/nqueen.rs
fn safe(c: int, cs: &Vec<int>) -> bool {
    let mut r1: int = 0;
    let r0: int = cs.len() as int;
    for &c1 in cs.iter() {
        if c1 == c || r0 - r1 == std::num::abs(c - c1) {
            return false;
        }
        r1 = r1 + 1;
    }
    return true;
}

fn solve(n: int, cs: Vec<int>) -> int {
    let x = cs.len() as int;
    if n == x {
        1
    } else {
        let mut s: int = 0;
        for c in range(0, n) {
            if safe(c, &cs) {
                let mut child: Vec<int> = cs.clone();
                child.push(c);
                s = s + solve(n, child)
            }
        }
        s
    }
}

fn main() {
    println!("{}", solve(12, Vec::new()))
}
